### PR TITLE
Graph4

### PR DESCRIPTION
**Pull request by Iamlovingit**
### change log
* add graph validation
* add splitter implementation

### CODE REVIEW

File: pkg/apis/serving/v1alpha1/inference_graph_validation.go [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+	return nil
+}
+
+//Validation of inference graph router type
+func validateInferenceGraphRouterType(ig *InferenceGraph) error {
+	nodes := ig.Spec.Nodes
+	for name, node := range nodes {
+		if node.RouterType != Single && node.RouterType != Splitter &&
+			node.RouterType != Ensemble && node.RouterType != Switch {
+			return fmt.Errorf("InferenceGraph[%s] Node[%s] RouterType[%s] is not supported, InferenceGraph supports RouterType List['Single', 'Spilitter', 'Ensemble', 'Switch']",
```
Comment (yuzisun) [inline pkg/apis/serving/v1alpha1/inference_graph_validation.go]:
```suggestion
			return fmt.Errorf("InferenceGraph[%s] Node[%s] RouterType[%s] is not supported, InferenceGraph supports RouterType List['Single', 'Splitter', 'Ensemble', 'Switch']",
```

File: pkg/apis/serving/v1alpha1/inference_graph_validation.go [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+		return fmt.Errorf(InvalidGraphNameFormatError, ig.Name, GraphNameFmt)
+	}
+	return nil
+}
+
+//Validation of inference graph router type
+func validateInferenceGraphRouterType(ig *InferenceGraph) error {
+	nodes := ig.Spec.Nodes
+	for name, node := range nodes {
+		if node.RouterType != Single && node.RouterType != Splitter &&
```
Comment (yuzisun) [inline pkg/apis/serving/v1alpha1/inference_graph_validation.go:93]: I think we can use kubebuilder enum annotation to validate this

File: pkg/apis/serving/v1alpha1/inference_graph_validation.go [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+		weight := 0
+		if node.RouterType == Splitter {
+			for _, route := range node.Routes {
+				if route.Weight == nil {
+					return fmt.Errorf("InferenceGraph[%s] Node[%s] Route[%s] missing the 'Weight'.", ig.Name, name, route.Service)
+				}
+				weight += int(*route.Weight)
+			}
+			if weight != 100 {
+				return fmt.Errorf("InferenceGraph[%s] Node[%s] splitter node all routing targets should be sum of 100", ig.Name, name)
```
Comment (yuzisun) [inline pkg/apis/serving/v1alpha1/inference_graph_validation.go]:
```suggestion
				return fmt.Errorf("InferenceGraph[%s] Node[%s] splitter node: the sum of traffic weights for all routing targets should be 100", ig.Name, name)
```

File: config/webhook/manifests.yaml [hunk 4]
```diff
@@ -108,3 +108,30 @@ webhooks:
...
+  - clientConfig:
+      caBundle: Cg==
+      service:
+        name: $(webhookServiceName)
+        namespace: $(kserveNamespace)
+        path: /validate-serving-kserve-io-v1alpha1-inferencegraph
+    failurePolicy: Fail
+    name: inferencegraph.kserve-webhook-server.validator
+    sideEffects: None
+    admissionReviewVersions: ["v1beta1"]
```
Comment (yuzisun) [inline config/webhook/manifests.yaml:127]: this should be v1?

File: manifests/charts/templates/webhookconfiguration.yaml [hunk 5]
```diff
@@ -107,3 +107,30 @@ webhooks:
...
+  - clientConfig:
+      caBundle: Cg==
+      service:
+        name: $(webhookServiceName)
+        namespace: $(kserveNamespace)
+        path: /validate-serving-kserve-io-v1alpha1-inferencegraph
+    failurePolicy: Fail
+    name: inferencegraph.kserve-webhook-server.validator
+    sideEffects: None
+    admissionReviewVersions: ["v1beta1"]
```
Comment (yuzisun) [inline manifests/charts/templates/webhookconfiguration.yaml:126]: same here

File: cmd/router/main.go [hunk 6]
```diff
@@ -27,11 +32,39 @@ func callService(serviceUrl string, input []byte, res chan<- []byte) error {
...
+	return nil
+}
+
 func routeStep(nodeName string, currentStep v1alpha1.InferenceRouter, graph v1alpha1.InferenceGraphSpec, input []byte, res chan<- []byte) error {
 	log.Printf("current step %v", nodeName)
 	response := map[string]interface{}{}
 	//For splitter and ABNTest call virtual service
 	if currentStep.RouterType == v1alpha1.Splitter {
+		route := pickupRoute(currentStep.Routes)
+		if route == nil {
```
Comment (yuzisun) [inline cmd/router/main.go:57]: in which case it could be nil ?

File: pkg/apis/serving/v1alpha1/inference_graph_validation.go (hunk 2 above)
Comment (Iamlovingit) [reply pkg/apis/serving/v1alpha1/inference_graph_validation.go:93]: Yes, it is better to use kubebuilder.

File: cmd/router/main.go (hunk 6 above)
Comment (Iamlovingit) [reply cmd/router/main.go:57]: as the logic, the `route` will never be `nil`, let me remove this check.

**Pull request merged**